File: Hard/3614_Process_String_with_Special_Operations_II/Solution.py
from typing import List, Optional


class Solution:
    def processStr(self, s: str, k: int) -> str:
        m = 0

        # Final length
        for c in s:
            if c == '*':
                if m > 0:
                    m -= 1
            elif c == '#':
                m *= 2
            elif c != '%':
                m += 1

        if k >= m:
            return '.'

        # Reverse simulation
        for c in reversed(s):
            if c == '*':
                m += 1

            elif c == '#':
                m //= 2
                if k >= m:
                    k -= m

            elif c == '%':
                k = m - 1 - k

            else:  # letter
                m -= 1
                if k == m:
                    return c

        return '.'


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    s1 = "a#b%*"
    k1 = 1
    print(solution.processStr(s1, k1))
    
    # test cases 2
    s2 = "cd%#*#"
    k2 = 3
    print(solution.processStr(s2, k2))

    # test cases 3
    s3 = "z*#"
    k3 = 0
    print(solution.processStr(s3, k3))
    
    
    

File: Hard/3614_Process_String_with_Special_Operations_II/Solution.java



public class Solution {
    public char processStr(String s, long k) {
        long m = 0;

        // Final length
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            if (c == '*') {
                if (m > 0) m--;
            } else if (c == '#') {
                m *= 2;
            } else if (c != '%') {
                m++;
            }
        }

        if (k >= m) return '.';

        // Reverse simulation
        for (int i = s.length() - 1; i >= 0; i--) {
            char c = s.charAt(i);

            if (c == '*') {
                m++;
            }
            else if (c == '#') {
                m /= 2;
                if (k >= m) k -= m;
            }
            else if (c == '%') {
                k = m - 1 - k;
            }
            else { // letter
                m--;
                if (k == m) return c;
            }
        }

        return '.';
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        System.out.println(solution.processStr("a#b%*", 1)); // Expected: 'a'
        
        // test cases 2
        System.out.println(solution.processStr("cd%#*#", 3)); // Expected: 'd'

        // test cases 3
        System.out.println(solution.processStr("z*#", 4)); // Expected: '.'
        
        
    }
}